syntax = "proto3";

import "google/protobuf/timestamp.proto";

service PostService {
  rpc CreatePost(CreatePostRequest) returns (CreatePostResponse) {}
  rpc UpdatePost(UpdatePostRequest) returns (UpdatePostResponse) {}
  rpc DeletePost(DeletePostRequest) returns (DeletePostResponse) {}
  rpc GetPost(GetPostRequest) returns (GetPostResponse) {}
  rpc GetPosts(GetPostsRequest) returns (GetPostsResponse) {}
}

message Post {
  uint64 Id = 1;
  uint64 AuthorId = 2;
  google.protobuf.Timestamp CreatedAt = 3;
  string Content = 4;
}

enum Error {
  PostNotFound = 0;
  PostNotOwned = 1;
}

message CreatePostRequest {
  uint64 AuthorId = 1;
  string Content = 2;
}

message CreatePostResponse {
  bool Success = 1;
  optional uint64 PostId = 2;
}

message UpdatePostRequest {
  uint64 PostId = 1;
  uint64 AuthorId = 2;
  string Content = 3;
}

message UpdatePostResponse {
  bool Success = 1;
  optional Error Error = 2;
}

message DeletePostRequest {
  uint64 PostId = 1;
  uint64 AuthorId = 2;
}

message DeletePostResponse {
  bool Success = 1;
  optional Error Error = 2;
}

message GetPostRequest {
  uint64 PostId = 1;
}

message GetPostResponse {
  bool Success = 1;
  optional Post Post = 2;
}

message GetPostsRequest {
  uint64 Page = 1;
  uint64 Size = 2;
}

message GetPostsResponse {
  repeated Post Posts = 1;
}
